### PR TITLE
CalculationExecutor

### PR DESCRIPTION
**Pull request by ObiWahn**

### CODE REVIEW

File: arangod/Aql/CalculationExecutor.cpp [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    //  AqlValue const& conditionResult = result->getValueReference(i, _conditionReg);
+
+    //  if (!conditionResult.toBoolean()) {
+    //    TRI_IF_FAILURE("CalculationBlock::executeExpressionWithCondition") {
+    //      THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
+    //    }
+    //    result->emplaceValue(i, _outReg,
+    //    arangodb::velocypack::Slice::nullSlice()); continue;
+    //  }
+    //}
```
Comment (hkernbach) [inline arangod/Aql/CalculationExecutor.cpp]: remove me

File: arangod/Aql/ExecutionNode.cpp [hunk 2]
```diff
@@ -1481,7 +1481,48 @@ void CalculationNode::toVelocyPackHelper(VPackBuilder& nodes, unsigned flags) co
...
+      this->expression(), std::move(expInVars)  // required by expression.execute
+      ,
+      std::move(expInRegs)  // required by expression.execute
+      ,
+      _conditionVariable  // unused for now (conditon reg?)
+  );
+
+  return std::make_unique<ExecutionBlockImpl<CalculationExecutor>>(&engine, this,
+                                                                   std::move(infos));
+  // return std::make_unique<CalculationBlock>(&engine, this);
```
Comment (hkernbach) [inline arangod/Aql/ExecutionNode.cpp]: remove?

File: arangod/Aql/ExecutorExpressionContext.cpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+      TRI_ASSERT(i < _regs.size());
+      if (doCopy) {
+        mustDestroy = true;  // as we are copying
+        return _inputRow.getValue((_regs)[i]).clone();
+      }
+      return _inputRow.getValue((_regs)[i]);
+    }
+  }
+
+  std::string msg("variable not found '");
```
Comment (hkernbach) [inline arangod/Aql/ExecutorExpressionContext.cpp]: remove?

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/2736/

File: arangod/Aql/CalculationExecutor.cpp [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+void doEvaluation(CalculationExecutorInfos& info, InputAqlItemRow& input,
+                  OutputAqlItemRow& output);
+void executeExpression(CalculationExecutorInfos& info, InputAqlItemRow& input,
+                       OutputAqlItemRow& output);
+}  // namespace _detail
+
+CalculationExecutorInfos::CalculationExecutorInfos(
+    RegisterId outputRegister, RegisterId nrInputRegisters,
+    RegisterId nrOutputRegisters, std::unordered_set<RegisterId> registersToClear
+
```
Comment (mchacki) [inline arangod/Aql/CalculationExecutor.cpp]: Formatting seems broken

File: arangod/Aql/CalculationExecutor.h [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+  Query* _query;
+  Expression* _expression;
+  std::vector<Variable const*> _expInVars;  // input variables for expresseion
+  std::vector<RegisterId> _expInRegs;       // input registers for expression?
+  bool _isReference = false;
+};
+
+/**
+ * @brief Implementation of Filter Node
```
Comment (mchacki) [inline arangod/Aql/CalculationExecutor.h]: Please fix the comment

File: arangod/Aql/CalculationExecutor.h [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+   *
+   * @return ExecutionState, and if successful exactly one new Row of AqlItems.
+   */
+  std::pair<ExecutionState, Stats> produceRow(OutputAqlItemRow& output);
+
+ public:
+  CalculationExecutorInfos& _infos;
+
+ private:
+  AqlValue getAqlValue(AqlValue const& inVarReg, size_t const& pos, bool& mustDestroy);
```
Comment (mchacki) [inline arangod/Aql/CalculationExecutor.h]: Private functions are not implemented / used, please remove

File: arangod/Aql/CalculationExecutor.h [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+  void initialize();
+
+ private:
+  Fetcher& _fetcher;
+
+  InputAqlItemRow _currentRow = InputAqlItemRow{CreateInvalidInputRowHint{}};
```
Comment (mchacki) [inline arangod/Aql/CalculationExecutor.h]: Please initialize in constructor, we do this everywhere else as well and only one place to look at

File: arangod/Aql/CalculationExecutor.h [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+  CalculationExecutorInfos(CalculationExecutorInfos const&) = delete;
+  ~CalculationExecutorInfos() = default;
+
+  RegisterId _outputRegisterId;
+
+  Query* _query;
+  Expression* _expression;
+  std::vector<Variable const*> _expInVars;  // input variables for expresseion
+  std::vector<RegisterId> _expInRegs;       // input registers for expression?
+  bool _isReference = false;
```
Comment (mchacki) [inline arangod/Aql/CalculationExecutor.h]: Please initialize in constructor, as in all other places. Than we only have single place to look at.

File: arangod/Aql/ExecutionNode.cpp [hunk 9]
```diff
@@ -1484,7 +1484,45 @@ void CalculationNode::toVelocyPackHelper(VPackBuilder& nodes, unsigned flags) co
...
+    TRI_ASSERT(infoIter->second.registerId < ExecutionNode::MaxRegisterId);
+
+    expInVars.emplace_back(var);
+    expInRegs.emplace_back(infoIter->second.registerId);
+  }
+
+  CalculationExecutorInfos infos(
+      outputRegister, getRegisterPlan()->nrRegs[previousNode->getDepth()],
+      getRegisterPlan()->nrRegs[getDepth()], getRegsToClear()
+
```
Comment (mchacki) [inline arangod/Aql/ExecutionNode.cpp]: Formatting seems to be entirely broken here.

File: arangod/Aql/ExecutorExpressionContext.h [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+  ExecutorExpressionContext(Query* query, InputAqlItemRow& inputRow,
+                            std::vector<Variable const*> const& vars,
+                            std::vector<RegisterId> const& regs)
+      : QueryExpressionContext(query), _inputRow(inputRow), _vars(vars), _regs(regs) {}
+
+  ~ExecutorExpressionContext() {}
+
+  size_t numRegisters() const override;
+
+  AqlValue const& getRegisterValue(size_t i) const override;
```
Comment (mchacki) [inline arangod/Aql/ExecutorExpressionContext.h]: please rename i => whatOver the semantics of the input is (e.g. RegisterID i cannot understand like this)

File: arangod/Aql/ExecutorExpressionContext.h [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+  Variable const* getVariable(size_t i) const override;
```
Comment (mchacki) [inline arangod/Aql/ExecutorExpressionContext.h]: Please rename i as well

File: arangod/Aql/ExecutorInfos.cpp [hunk 12]
```diff
@@ -25,12 +25,12 @@
 using namespace arangodb::aql;
 
 ExecutorInfos::ExecutorInfos(
-    std::shared_ptr<std::unordered_set<RegisterId>> inputRegisters,
-    std::shared_ptr<std::unordered_set<RegisterId>> outputRegisters,
+    std::shared_ptr<std::unordered_set<RegisterId>> readableInputRegisters,
```
Comment (mchacki) [inline arangod/Aql/ExecutorInfos.cpp]: isn't readableInput doubled ?
input := read
output := write

File: arangod/Aql/Stats.h [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+#ifndef ARANGOD_AQL_STATS_H
+#define ARANGOD_AQL_STATS_H 1
+
+#include "Aql/ExecutionStats.h"
+
+namespace arangodb {
+namespace aql {
+
+// no-op statistics for all Executors that don't have custom stats.
+class NoStats {};
```
Comment (mchacki) [inline arangod/Aql/Stats.h]: Good file, should contain all Stat Types

File: tests/Aql/CalculationExecutorTest.cpp [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+          std::tie(state, stats) = testee.produceRow(result);
+          REQUIRE(state == ExecutionState::DONE);
+          REQUIRE(!result.produced());
+        }
+      }
+    }
+
+  }  // GIVEN
+
+  GIVEN("there are rows in the upstream, and the last one has to be filtered") {
```
Comment (mchacki) [inline tests/Aql/CalculationExecutorTest.cpp]: please rename test properly

File: arangod/Aql/ExecutorInfos.cpp (hunk 12 above)
Comment (ObiWahn) [reply arangod/Aql/ExecutorInfos.cpp]: This the the solution @goedderz and I came up with after some discussion. It seemed to be the most correct way. As there are are for example input registers that are not readable. They just get copied as they might be in registers to keep.

**Pull request merged**